openedx/core/djangoapps/programs/utils.py

```diff
@@ -461,57 +462,70 @@ def _attach_course_run_upgrade_url(self, run_mode):
     def _attach_course_run_may_certify(self, run_mode):
         run_mode['may_certify'] = self.course_overview.may_certify()
 
-    def _check_enrollment_for_user(self, course_run):
-        applicable_seat_types = self.data['applicable_seat_types']
-
-        (enrollment_mode, active) = CourseEnrollment.enrollment_mode_for_user(
-            self.user,
-            CourseKey.from_string(course_run['key'])
+    def _filter_out_courses_with_entitlements(self, courses):
+        course_uuids = set(course['uuid'] for course in courses)
+        # Filter the entitlements' modes with a case-insensitive match against applicable seat_types
+        query = Q()
+        for seat_type in self.data['applicable_seat_types']:
+            query = query | Q(mode__iexact=seat_type)
+        query = Q(course_uuid__in=course_uuids) & query
+        entitlements = self.user.courseentitlement_set.filter(query)
+        courses_with_entitlements = set(str(entitlement.course_uuid) for entitlement in entitlements)
+        return [course for course in courses if course['uuid'] not in courses_with_entitlements]
```

Comment:
We need to test that this will handle the mode no-id-professional (I need to check the string for this) will match a professional applicable seat type. As there is not a no-id-professional applicable seat type, it is just a variation of professional.

Reply:
👍 I will add a test for that

Reply:
We should explicitly cast `course['uuid']` to `str` or `unicode` to match what is expected for `courses_with_entitlements`. Even though `course[`uuid`]` is returned as a `str` now, it may not in the future, so it would be better to be explicit when making comparisons.

Reply:
I believe `course['uuid'] should always be a `str` because it's a value in the program data json object.